Remove RewindBlockIndex logic

dhruv opened the pull request:
Closes #17862

Context from [original comment](https://github.com/bitcoin/bitcoin/issues/17862#issuecomment-744285188) (minor edits):

`RewindBlockIndex()` is a mechanism to allow nodes to be upgraded after segwit activation, while still keeping their chainstate/datadir in a consistent state. It works as follows:

- A pre-segwit (i.e. v0.13.0 or older) node is running.
-  Segwit activates. The pre-segwit node remains sync'ed to the tip, but is not enforcing the new segwit rules.
- The user upgrades the node to a segwit-aware version (v0.13.1 or newer).
- On startup, in `AppInitMain()`, `RewindBlockIndex()` is called. This walks the chain backwards from the tip, disconnecting and erasing blocks that from after segwit activation that weren't validated with segwit rules.
- those blocks are then redownloaded (with witness data) and validated with segwit rules.

This logic probably isn't required any more since:

- Segwit activated at height 481824, when the block chain was 130GB and the total number of txs was 250 million. Today, we're at height 667704, the blockchain is over 315GB and the total number of txs is over 600 million. Even if 20% of that added data is witness data (a high estimate), then around 150GB of transactions would need to be rewound to get back to segwit activation height. It'd probably be faster to simply validate from genesis, especially since we won't be validating any scripts before the assumevalid block. It's also unclear whether rewinding 150GB of transactions would even work. It's certainly never been tested.
- Bitcoin Core v0.13 is hardly used any more. https://luke.dashjr.org/programs/bitcoin/files/charts/software.html shows less than 50 nodes running it. The software was EOL on Aug 1st 2018. It's very unlikely that anyone is running 0.13 and will want to upgrade to 0.22.

This PR introduces `NeedsRedownload()` which merely checks for insufficiently validated segwit blocks and requests that the user restarts the node with `-reindex`. Reindexing the block files upon restart will make the node rebuild chain state and block index from the `blk*.dat` files on disk. The node won't be able to index the blocks with `BLOCK_OPT_WITNESS`, so they will be missing from the chain and be re-downloaded, with witness data.

Removing this code allows the following (done in follow-up #21090):

- removal of tests using `segwitheight=-1` in `p2p_segwit.py`.
- in turn, that allows us to drop support for `-segwitheight=-1`, which is only supported for that test.
- that allows us to always set `NODE_WITNESS` in our local services. The only reason we don't do that is to support `-segwitheight=-1`.
- that in turn allows us to drop all of the `GetLocalServices() & NODE_WITNESS` checks inside `net_processing.cpp`, since our local services would always include `NODE_WITNESS`

dhruv committed:
[validation] RewindBlockIndex no longer needed

Instead of rewinding blocks, we request that the user restarts with
-reindex

jnewbery commented:
Concept ACK

dhruv commented:
The linter was failing on all pull requests with the same error when I pushed so it is likely unrelated to the commits here and will get resolved with the next push. Ready for review.

practicalswift commented:
@dhruv Rebase on `master` and the `boost/thread/mutex.hpp` warning will go away: it was fixed in #21010 :)

dhruv commented:
Rebased against master and linter is passing. Thanks, @practicalswift.

theStack commented:
Concept ACK

ccdle12 reviewed (commented):
Concept ACK

dhruv commented on src/validation.cpp:
```diff
@@ -4381,143 +4381,19 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
-                return false;
+        for (CBlockIndex* block = m_chain.Genesis(); block != nullptr; block = m_chain.Next(block)) {
+            if (IsWitnessEnabled(block->pprev, params.GetConsensus()) && !(block->nStatus & BLOCK_OPT_WITNESS)) {
+                // block is insuficiently validated for a segwit client
```
Done.

dhruv commented on src/init.cpp:
```diff
@@ -1821,11 +1818,7 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
-        // The option to not set NODE_WITNESS is only used in the tests and should be removed.
-        nLocalServices = ServiceFlags(nLocalServices | NODE_WITNESS);
-    }
+    nLocalServices = ServiceFlags(nLocalServices | NODE_WITNESS);
```
That makes sense. Thanks!

dhruv commented:
Thanks for the review @ccdle12. Comments addressed. Ready for further review.

luke-jr commented:
Makes sense, if the rationale was simply to ensure the block gets redownloaded.

If we focus on validation, however, we would want this around for Taproot. But I'm not sure that's what its purpose is.

Re-validating blocks on upgrade seems like a feature we don't have today, and should be implemented separately from this (without redownloading).

(Therefore, Concept ACK)

dhruv commented:
@luke-jr You're right, the removed code erases insufficiently validated blocks (which do not have witness data and can't be properly validated by a segwit-aware node) and re-downloads them. AFAICT, with Taproot, the post-activation blocks will have witness data, so we'll need to implement a different function to re-validate the insufficiently validated blocks after upgrade.

ariard commented:
Concept ACK.

jnewbery commented on src/init.cpp:
```diff
@@ -1687,26 +1687,23 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
-            bool failed_rewind{false};
-            // Can't hold cs_main while calling RewindBlockIndex, so retrieve the relevant
+            bool needs_ibd{false};
+            // Can't hold cs_main while calling NeedsIBD, so retrieve the relevant
```
This isn't true. `NeedsIBD()` grabs cs_main and holds throughout. You could change it to requires cs_main, then run this entire block under one cs_main lock.

dhruv replied:
Done.

jnewbery commented on src/validation.h:
```diff
@@ -673,7 +673,7 @@ class CChainState
     /** Replay blocks that aren't fully applied to the database. */
     bool ReplayBlocks(const CChainParams& params);
-    bool RewindBlockIndex(const CChainParams& params) LOCKS_EXCLUDED(cs_main);
+    bool NeedsIBD(const CChainParams& params) LOCKS_EXCLUDED(cs_main);
```
Make `[nodiscard]`?

dhruv replied:
Great idea. Done.

jnewbery commented on src/validation.cpp:
```diff
@@ -4381,143 +4381,19 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
-    }
-
     {
         LOCK(cs_main);
```
Perhaps just annotate this function as requiring cs_main and assert that it's held.

dhruv replied:
Done.

jnewbery commented on src/validation.cpp:
```diff
@@ -4381,143 +4381,19 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
-            if (!FlushStateToDisk(params, state, FlushStateMode::ALWAYS)) {
-                LogPrintf("RewindBlockIndex: unable to flush state to disk (%s)\n", state.ToString());
-                return false;
+        for (CBlockIndex* block = m_chain.Genesis(); block != nullptr; block = m_chain.Next(block)) {
```
Can you use a range based for loop since you're iterating over all members of this container?

dhruv replied:
I took @ariard's suggestion and started from the `params.SegwitHeight()` - also changed it to a while loop that reads cleaner.

Out of curiosity, if a defined `[]` operator all that's needed to use a range-based for loop?

ajtowns replied:
range-based for loops need an `x.begin()` and `x.end()` (or `begin(x)` and `end(x)`) that return iterators, and the iterators need to accept `++it` and `*it` and be comparable with `!=` -- https://en.cppreference.com/w/cpp/language/range-for

jnewbery replied:
My mistake. For some reason I thought m_chain was a container here.

jnewbery reviewed (commented):
Concept ACK. I think it might make sense to split this into two PRs. The second commit is quite large and makes a lot of changes.

ariard commented on src/chainparams.cpp:
```diff
@@ -461,11 +461,8 @@ void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
     if (args.IsArgSet("-segwitheight")) {
         int64_t height = args.GetArg("-segwitheight", consensus.SegwitHeight);
-        if (height < -1 || height >= std::numeric_limits<int>::max()) {
+        if (height < 0 || height >= std::numeric_limits<int>::max()) {
```
Can you commit-split the changes around `segwithheight` from the ones around `NODE_WITNESS` ? Better to review any changes in net_processing on their own.

dhruv replied:
Done - moved the second commit to #21090 and split it up into several there.

ariard commented on src/net_processing.cpp:
```diff
@@ -1948,7 +1948,7 @@ void static ProcessGetData(CNode& pfrom, Peer& peer, const CChainParams& chainpa
 static uint32_t GetFetchFlags(const CNode& pfrom) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     uint32_t nFetchFlags = 0;
-    if ((pfrom.GetLocalServices() & NODE_WITNESS) && State(pfrom.GetId())->fHaveWitness) {
+    if (State(pfrom.GetId())->fHaveWitness) {
```
I think this is buggy. Always setting `NODE_WITNESS` to our local service flags doesn't mean we will never meet peers which aren't advertising it. Of course it's unlikely there is that much of those peers deployed but otherwise I believe we'll wrongly unserialize received txn.

jnewbery replied:
Why do you think this is buggy? We'll unserialize using `MSG_WITNESS_FLAG` iff `fHaveWitness` is set to true, which happens if the peer includes `NODE_WITNESS` in its version message.

ariard replied:
I think we unserialize if `fAllowWitness=true`, `MSG_WITNESS_FLAG` is used by a getdata sender to require serialization of witnesses but not at unserialization itself?

That said, you're right in the sense that `fHaveWitness` implies they advertise their local services with `NODE_WITNESS`. And those ones are static so it should be good. Further, the conditional could be already reduced to `fHaveWitness` check only, a test node with `-segwitheight== std::numeric_limits<int>::max()` will have `fHaveWitness=false` ?

jnewbery replied:
> MSG_WITNESS_FLAG is used by a getdata sender to require serialization of witnesses but not at unserialization itself

Sorry, yes - I wrote "unserialize" above when I meant "serialize"

> Further, the conditional could be already reduced to fHaveWitness check only, a test node with -segwitheight== std::numeric_limits<int>::max() will have fHaveWitness=false ?

I don't think that's right. Before this PR, it's checking that both:

- we have signaled NODE_WITNESS to the peer
- the peer has signaled NODE_WITNESS to us

we can only use witness serialization on the connection when both of those conditions are true.

After this PR, we unconditionally signal NODE_WITNESS to all peers, so we only need to check that they signaled NODE_WITNESS to us (which is what is indicated by fHaveWitness).

ariard replied:
Gotcha, reading comment L869 in `src/net_processing.cpp` : "Note that pnode->GetLocalServices() is a reflection of the local services we were offering when the CNode object was created for this peer" makes me understood this is _our_ local services, not _their_. Renaming the method `GetAnnouncedToPeerServices` or something similar would be clearer.

But further, if our local services are static beyond initialization does it make sense to track them per-peer, the same flags should be announced to all our peers during a runtime lifecycle ?

> we can only use witness serialization on the connection when both of those conditions are true.

Does BIP144 really require double opt-in to fetch a tx and its witnesses with a getdata ? Not clear reading the BIP no more the getdata processing code.

```
            // WTX and WITNESS_TX imply we serialize with witness
            int nSendFlags = (inv.IsMsgTx() ? SERIALIZE_TRANSACTION_NO_WITNESS : 0);
            connman.PushMessage(&pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *tx));
```

jnewbery replied:
I'm sure there are lots of improvements that could be made around how we track services (and indeed will be made as part of https://github.com/bitcoin/bitcoin/issues/19398). However, that seems orthogonal to this PR. Here, we just want to ensure that mainnet behaviour is unaffected.

ariard commented on src/net_processing.cpp:
```diff
@@ -2717,7 +2716,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         uint64_t nCMPCTBLOCKVersion = 0;
         vRecv >> fAnnounceUsingCMPCTBLOCK >> nCMPCTBLOCKVersion;
-        if (nCMPCTBLOCKVersion == 1 || ((pfrom.GetLocalServices() & NODE_WITNESS) && nCMPCTBLOCKVersion == 2)) {
+        if (nCMPCTBLOCKVersion == 1 || nCMPCTBLOCKVersion == 2) {
```
Maybe better to defer those change to #20799, which will achieve the same IIRC but can be reasoned on their own ?

dhruv replied:
This diff is now moved to #21090 but I left it in because the check is redundant if this code is merged. IIUC, #20799 eliminates the entire conditional so they can both be reasoned for independently?

ariard reviewed (commented):
> If we focus on validation, however, we would want this around for Taproot. But I'm not sure that's what its purpose is.

I don't think we need it for Taproot as it doesn't introduce new consensus data but extend meaning of already existing witness ?

We might need again that kind of logic in the future when we do have a soft-fork introducing new consensus data but AFAIK there is no such studied proposal. And if this happen I hope we can be smarter to avoid re-downloading.

dhruv commented on src/validation.cpp:
```diff
@@ -4381,143 +4381,19 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
-                LogPrintf("RewindBlockIndex: unable to flush state to disk (%s)\n", state.ToString());
-                return false;
+        for (CBlockIndex* block = m_chain.Genesis(); block != nullptr; block = m_chain.Next(block)) {
+            if (IsWitnessEnabled(block->pprev, params.GetConsensus()) && !(block->nStatus & BLOCK_OPT_WITNESS)) {
```
Great idea. Updated to start from `params.SegwitHeight()`

dhruv commented:
Comments addressed. The second commit has been broken out into #21090. Ready for further review.

jnewbery commented on src/init.cpp:
```diff
@@ -1686,26 +1686,24 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+                LOCK(cs_main);
+                for (CChainState* chainstate : chainman.GetAll()) {
+                    if (!fReset) {
+                        uiInterface.InitMessage(_("Checking if IBD is needed...").translated);
```
This will print the message twice if there are two `CChainState`s.

dhruv replied:
Done.

jnewbery commented on src/init.cpp:
```diff
@@ -1686,26 +1686,24 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
-                        failed_rewind = true;
-                        break; // out of the per-chainstate loop
+            bool needs_ibd{false};
+            {
```
We can avoid some deep nesting and the need for a temporary variable to pass the result of the predicate out of the loop by using an stl algorithm:

```diff
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1686,27 +1686,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 break;
             }
 
-            bool needs_ibd{false};
-            {
-                LOCK(cs_main);
-                for (CChainState* chainstate : chainman.GetAll()) {
-                    if (!fReset) {
-                        uiInterface.InitMessage(_("Checking if IBD is needed...").translated);
-                        if (chainstate->NeedsIBD(chainparams)) {
-                            strLoadError = _(
-                                "Segwit blocks are insufficiently validated. "
-                                "Please delete blocks dir and chain state dir and restart.");
-                            needs_ibd = true;
-                            break; // out of the per-chainstate loop
-                        }
-                    }
+            if (!fReset) {
+                uiInterface.InitMessage(_("Checking if redownload is needed...").translated);
+                LOCK(cs_main);
+                auto chainstates{chainman.GetAll()};
+                if (std::any_of(chainstates.begin(), chainstates.end(),
+                                [&chainparams] (CChainState* cs) {return cs->NeedsIBD(chainparams);})) {
+                    strLoadError = _(
+                        "Segwit blocks are insufficiently validated. "
+                        "Please delete blocks dir and chain state dir and restart.");
+                    break;
                 }
             }
 
-            if (needs_ibd) {
-                break; // out of the chainstate activation do-while
-            }
-
             bool failed_verification = false;
```

This is exactly what `std::any_of` is designed for. Instead of creating a boolean, manually iterating through a range, executing a predicate on each member and then setting the boolean and breaking out if the predicate returns true, we can let the standard library do the heavy lifting. That makes the code more expressive by raising the level of abstraction and communicating what we want the code to do, and not what the individual steps to do it are.

dhruv replied:
Beautiful change. Thank you for teaching so patiently!

jnewbery replied:
Ooops. I guess the lambda needs an annotation too. Something like this maybe:

```
                if (std::any_of(chainstates.begin(), chainstates.end(),
                                [&chainparams] (CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {return cs->NeedsIBD(chainparams);})) {
```

dhruv replied:
That worked. For my education, what happened there and how did you know the lambda annotation would help?

I suspect the annotation is evaluated with the limited scope of the bound variables, so it does not "know" about the locks that are acquired in the outer scope. Is that even close?

jnewbery replied:
As discussed, there was an error message about `NeedsIBD()` requiring cs_main. Since the compiler does not know that when it's inside the lambda, it's in a scope that already has cs_main, we need to annotate it.

jnewbery commented on src/validation.cpp:
```diff
@@ -4390,143 +4390,22 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
 
-//! Helper for CChainState::RewindBlockIndex
-void CChainState::EraseBlockData(CBlockIndex* index)
+bool CChainState::NeedsIBD(const CChainParams& params)
```
Consider renaming this function `NeedsRedownload`. All nodes that are starting fresh or restarting after some time need IBD. This function is specifically for nodes that need to delete blocks and redownload them.

dhruv replied:
Makes sense. Done.

jnewbery commented on test/functional/p2p_segwit.py:
```diff
@@ -1956,21 +1959,49 @@ def test_non_standard_witness(self):
         assert softfork_active(self.nodes[2], 'segwit')
 
-        # Make sure this peer's blocks match those of node0.
+        # Make sure all peers have the same blocks.
```
This loop seems useless. `sync_blocks()` ensures that the nodes have the same block hash at their tip. We assume that if the block hash at the tip is the same then all blocks in the chain are the same (if not, then bitcoin is fundamentally broken).

dhruv replied:
Fair enough. I didn't remember that `sync_blocks()` verifies that the best hashes are equal. Done.

dhruv commented:
Comments addressed. Rebased. Ready for further review.

jnewbery commented:
Code review ACK 0af05b95e9c829ff581bb76a0998b5a90386e27a

ariard commented on src/init.cpp:
```diff
@@ -1686,27 +1686,18 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+                                [&chainparams](CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return cs->NeedsRedownload(chainparams); })) {
+                    strLoadError = _(
+                        "Segwit blocks are insufficiently validated. "
+                        "Please delete blocks dir and chain state dir and restart.");
```
Should we be more cautious with directly summoning users to delete directories ?

Maybe we should give a more detailed path (`.bitcoin/blocks`, `.bitcoin/chainstate`) but not sure if it's platform dependent. I just want to minimize the risk of someone deleting by mistake its wallet.

dhruv replied:
Good idea! Done.

ariard reviewed (commented):
Code Review ACK 0af05b95 pending on comment about error message.

dhruv commented:
Updated to log the blocksdir and chainstate dir to the user so they do not accidentally delete their wallet. Ready for further review.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1686,27 +1686,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+                    auto chainstate_dir = GetDataDir() / CHAINSTATE_DIRNAME;
+                    strLoadError = strprintf(_(
+                        "Segwit blocks are insufficiently validated. "
+                        "Please delete blocks dir(%s) and chainstate dir(%s) and restart."), GetBlocksDir(), chainstate_dir);
```
A `-reindex` should do the same with less effort?

dhruv replied:
When a pre-segwit node upgrades to segwit, it has no way to sufficiently validate to the "new" consensus rules from the blocks on disk since witness data was never relayed to it. Help for `-reindex` says "Rebuild chain state and block index from the blk*.dat files on disk". Looking at the code also does not suggest to me that `-reindex` would cause the node to re-download.

Am I missing something?

jnewbery replied:
Ah, I think this is true. `-reindex` will cause the node to try to rebuild the chain state from the block files, which will work only as far as the segwit activation height if the blocks after that are serialized without witness. The node will then try to sync by downloading all blocks from peers as normal. The test passes with the following change:

```diff
diff --git a/test/functional/p2p_segwit.py b/test/functional/p2p_segwit.py
index 219c15f673..76d8967a41 100755
--- a/test/functional/p2p_segwit.py
+++ b/test/functional/p2p_segwit.py
@@ -1973,15 +1973,8 @@ class SegWitTest(BitcoinTestFramework):
         with self.nodes[2].assert_debug_log(expected_msgs=["Segwit blocks are insufficiently validated. Please delete blocks dir and chain state dir and restart."], timeout=10):
             self.nodes[2].start(["-segwitheight={}".format(SEGWIT_HEIGHT)])
 
-        # As directed, the user deletes the blocks and chainstate directories and restarts the node
-        datadir = get_datadir_path(self.options.tmpdir, 2)
-        blocks_dir = os.path.join(datadir, "regtest", "blocks")
-        chainstate_dir = os.path.join(datadir, "regtest", "chainstate")
-        shutil.rmtree(blocks_dir)
-        shutil.rmtree(chainstate_dir)
-
-        self.start_node(2, extra_args=["-segwitheight={}".format(SEGWIT_HEIGHT)])
-        assert_equal(self.nodes[2].getblockcount(), 0)
+        self.start_node(2, extra_args=["-reindex=1", "-segwitheight={}".format(SEGWIT_HEIGHT)])
+        assert_equal(self.nodes[2].getblockcount(), SEGWIT_HEIGHT - 1)
         self.connect_nodes(0, 2)
 
         # We reconnect more than 100 blocks, give it plenty of time
```

There are a bunch of `AcceptBlock` errors since the blocks after height SEGWIT_HEIGHT aren't serialized properly:

```
node2 2021-02-16T10:05:38.943607Z [loadblk] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size 
 node2 2021-02-16T10:05:38.943663Z [loadblk] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size 
 node2 2021-02-16T10:05:38.943712Z [loadblk] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size 
[...]
```

and when the node connects to its peers, it downloads and syncs to the tip, validating segwit rules as it goes.

One downside is that the old blocks are retained on disk and never deleted. However, that was the case before this PR as well.

dhruv replied:
Ah, what I wasn't thinking about is that it is normal for the block _index_ to not be a 1:1 representation of the blk*.dat files. It happens all the time with reorgs, etc. So, upon `-reindex` node won't be able to validate `[SEGWIT_HEIGHT, consensusTip]` blocks, will have `tip == SEGWIT_HEIGHT - 1`. It will then proceed to sync headers and therefore re-download `[SEGWIT_HEIGHT, consensusTip]`.

Thank you for the great suggestion @MarcoFalke! Makes for much simpler code. Updated.

dhruv commented:
Force pushed last night to fix failing test.

ariard commented:
Code Review ACK d9370c1, thanks for taking the suggestion, Marco one sounds fine to me if you want to take it.

jnewbery commented on src/validation.h:
```diff
@@ -556,7 +557,7 @@ class CChainState
         bool in_memory,
         bool should_wipe,
-        std::string leveldb_name = "chainstate");
+        std::string leveldb_name = CHAINSTATE_DIRNAME);
```
Why is this change needed?

dhruv replied:
You are right - it is no longer needed with the `-reindex` change. Removed.

dhruv commented:
Force pushed to use `-reindex` instead of asking the user to delete blocks dir and chainstatedir. Ready for further review!

jnewbery commented on test/functional/p2p_segwit.py:
```diff
@@ -1956,23 +1956,38 @@ def test_non_standard_witness(self):
+        self.stop_node(2)
+        with self.nodes[2].assert_debug_log(expected_msgs=[
+                'Segwit blocks are insufficiently validated. Please restart with -reindex.'], timeout=10):
+            self.nodes[2].start(["-segwitheight={}".format(SEGWIT_HEIGHT)])
```
(please don't change the branch for this, but if you need to retouch for some other reason) you can use an f-string here:

```suggestion
            self.nodes[2].start([f"-segwitheight={SEGWIT_HEIGHT}"])
```

f-strings are supported from python 3.6, and have a slightly cleaner syntax.

You can also use an f-string in the `extra_args` argument below.

dhruv replied:
Thanks for showing me that. Done.

jnewbery reviewed (commented):
utACK 5e818578524ee8419bcaf610cf1394817d867663

The CI failure seemed unrelated. I've restarted it.

ajtowns commented on src/validation.h:
```diff
@@ -684,7 +684,7 @@ class CChainState
     /** Replay blocks that aren't fully applied to the database. */
     bool ReplayBlocks(const CChainParams& params);
-    bool RewindBlockIndex(const CChainParams& params) LOCKS_EXCLUDED(cs_main);
+    [[nodiscard]] bool NeedsRedownload(const CChainParams& params) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Add a doxygen comment? `/** Check if chain state needs to be redownloaded due to lack of witness data */` ?

Should be a `const` method, as far as I can see.

dhruv replied:
Done and done.

ajtowns reviewed (commented):
Here's some suggested changes. Looks good to me though.

dhruv commented on src/init.cpp:
```diff
@@ -1686,27 +1686,16 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+                auto chainstates{chainman.GetAll()};
+                if (std::any_of(chainstates.begin(), chainstates.end(),
+                                [&chainparams](CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return cs->NeedsRedownload(chainparams); })) {
+                    strLoadError = _("Segwit blocks are insufficiently validated. Please restart with -reindex.");
```
Done.

dhruv commented on src/validation.cpp:
```diff
@@ -4390,143 +4390,22 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
-        }
-    }
+    // At and above params.SegwitHeight, segwit consensus rules must be validated
+    CBlockIndex* block = m_chain[params.GetConsensus().SegwitHeight];
```
Yeah, it's a little more intuitive to go backwards. It's also faster. Updated.

Note that the code was not checking each block in `[segwit_height, tip]`. It was merely looking for one insufficiently validated block and returning true. It continues to do the same thing which is why going backwards from tip is faster.

ajtowns replied:
In the normal case it checks every block in `[segwit_height, tip]` and returns false because it didn't find a block. :) (Unless I'm very confused?)

dhruv replied:
You're correct. I misunderstood what you were saying.

dhruv commented on src/init.cpp:
```diff
@@ -1686,27 +1686,16 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
-            if (failed_rewind) {
-                break; // out of the chainstate activation do-while
+            if (!fReset) {
+                uiInterface.InitMessage(_("Checking if redownload is needed...").translated);
```
Nice idea. Removed.

dhruv commented:
Thanks for the reviews @ajtowns and @jnewbery. My apologies on the delay. Our twins arrived a few days ago so we are still finding our new routine :)

Comments addressed. Ready for further review.

jnewbery commented:
utACK e750a774b8375dcbab9b804e2fd1334f1f62d6e4

jnewbery commented:
Your latest rebase has reintroduced the `EraseBlockData()` declaration in validation.h L776.

dhruv commented:
Ah, yes. I was just fixing it while you noticed as well. Thanks, @jnewbery . Ready for further review.

jnewbery commented:
utACK 1aecaac8b46a936634245739efff852f03c32b55

ariard commented:
Code Review ACK 1aecaac

jonatack commented on src/validation.cpp:
```diff
@@ -4431,143 +4431,22 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
-                LogPrintf("RewindBlockIndex: unable to flush state to disk (%s)\n", state.ToString());
-                return false;
-            }
+    while (block != nullptr && block->nHeight >= params.GetConsensus().SegwitHeight) {
```
```diff
-    CBlockIndex* block = m_chain.Tip();
+    CBlockIndex* block{m_chain.Tip()};
+    const int segwit_height{params.GetConsensus().SegwitHeight};
 
-    while (block != nullptr && block->nHeight >= params.GetConsensus().SegwitHeight) {
+    while (block != nullptr && block->nHeight >= segwit_height) {
```

dhruv replied:
Done

jonatack commented on test/functional/p2p_segwit.py:
```diff
@@ -1956,23 +1956,38 @@ def test_non_standard_witness(self):
+        assert SEGWIT_HEIGHT < self.nodes[2].getblockcount()
+        assert softfork_active(self.nodes[0], 'segwit')
+        assert softfork_active(self.nodes[1], 'segwit')
+        assert 'segwit' not in self.nodes[2].getblockchaininfo()['softforks']
```
```diff
         # All nodes are caught up and node 2 is a pre-segwit node that will soon upgrade.
+        for n in range(2):
+            assert_equal(self.nodes[n].getblockcount(), self.nodes[2].getblockcount())
+            assert softfork_active(self.nodes[n], "segwit")
         assert_equal(self.nodes[0].getblockcount(), self.nodes[2].getblockcount())
-        assert_equal(self.nodes[1].getblockcount(), self.nodes[2].getblockcount())
         assert SEGWIT_HEIGHT < self.nodes[2].getblockcount()
-        assert softfork_active(self.nodes[0], 'segwit')
-        assert softfork_active(self.nodes[1], 'segwit')
         assert 'segwit' not in self.nodes[2].getblockchaininfo()['softforks']
```

dhruv replied:
Done

jonatack commented on test/functional/p2p_segwit.py:
```diff
@@ -1956,23 +1956,38 @@ def test_non_standard_witness(self):
+        assert_equal(self.nodes[1].getblockcount(), self.nodes[2].getblockcount())
+        assert softfork_active(self.nodes[0], 'segwit')
+        assert softfork_active(self.nodes[1], 'segwit')
         assert softfork_active(self.nodes[2], 'segwit')
```
```diff
         # The upgraded node syncs headers and performs redownload
-        assert_equal(self.nodes[0].getblockcount(), self.nodes[2].getblockcount())
-        assert_equal(self.nodes[1].getblockcount(), self.nodes[2].getblockcount())
-        assert softfork_active(self.nodes[0], 'segwit')
-        assert softfork_active(self.nodes[1], 'segwit')
-        assert softfork_active(self.nodes[2], 'segwit')
+        for n in range(2):
+            assert_equal(self.nodes[n].getblockcount(), self.nodes[2].getblockcount())
+        for n in range(3):
+            assert softfork_active(self.nodes[n], "segwit")
```

dhruv replied:
Done

jonatack commented on src/validation.h:
```diff
@@ -724,7 +724,9 @@ class CChainState
-    bool RewindBlockIndex(const CChainParams& params) LOCKS_EXCLUDED(cs_main);
+
+    /** Check if chain state needs to be redownloaded due to lack of witness data */
+    [[nodiscard]] bool NeedsRedownload(const CChainParams& params) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
```diff
-    /** Check if chain state needs to be redownloaded due to lack of witness data */
-    [[nodiscard]] bool NeedsRedownload(const CChainParams& params) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    /** Whether the chain state needs to be reindexed due to lack of witness data. */
+    [[nodiscard]] bool NeedsReindex(const CChainParams& params) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
     bool LoadGenesisBlock(const CChainParams& chainparams);
```

dhruv replied:
I left it as `NeedsRedownload` as that represents more accurately, what's needed. Restarting with `-reindex` is just the suggestion we make to the user that accomplishes the redownload. To me, `NeedsReindex` suggests merely rebuilding an index from what we have would be enough.

Updated the comment though.

jonatack reviewed (commented):
Tested Approach ACK, a few non-blocking thoughts below

jonatack commented on src/init.cpp:
```diff
@@ -1697,27 +1697,16 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
-                    }
-                }
-            }
 
```
(nit, remove extra blank line)

dhruv replied:
Done

dhruv commented:
Thank you for the review @jonatack. Comments addressed. Ready for further review.

jonatack commented:
ACK 644827722b9eba8af40576505a3d5444272f29c3 per `git range-diff e0bc27a 1aecaac 6448277` and debug build/ran bitcoind

Small suggestion to save a few `getblockcount()` calls, happy to re-ACK if you update. Sorry for not noticing on the first pass:

<details><summary>suggestion</summary><p>

```diff
         # All nodes are caught up and node 2 is a pre-segwit node that will soon upgrade.
+        node_2_height = self.nodes[2].getblockcount()
         for n in range(2):
-            assert_equal(self.nodes[n].getblockcount(), self.nodes[2].getblockcount())
+            assert_equal(self.nodes[n].getblockcount(), node_2_height)
             assert softfork_active(self.nodes[n], "segwit")
-        assert SEGWIT_HEIGHT < self.nodes[2].getblockcount()
+        assert SEGWIT_HEIGHT < node_2_height
         assert 'segwit' not in self.nodes[2].getblockchaininfo()['softforks']
 
         # Restarting node 2 should result in a shutdown because the blockchain consists of
@@ -1981,8 +1982,9 @@ class SegWitTest(BitcoinTestFramework):
         self.sync_blocks(timeout=240)
 
         # The upgraded node syncs headers and performs redownload
+        node_2_height = self.nodes[2].getblockcount()
         for n in range(2):
-            assert_equal(self.nodes[n].getblockcount(), self.nodes[2].getblockcount())
+            assert_equal(self.nodes[n].getblockcount(), node_2_height)
```
</p></details>

jnewbery reviewed (commented):
utACK 644827722b9eba8af40576505a3d5444272f29c3

dhruv commented:
> Small suggestion to save a few getblockcount() calls, happy to re-ACK if you update. Sorry for not noticing on the first pass:

Thanks, @jonatack. I will update it if I end up rebasing or addressing other comments.

jamesob commented:
Concept ACK, will review soon.

jnewbery commented:
@ariard do you mind re-reviewing this? I believe all of your review comments are addressed.

glozow reviewed (commented):
Concept ACK to removing code that's no longer necessary and the simplifications it allows

jnewbery commented on src/init.cpp:
```diff
@@ -1698,29 +1698,17 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+                LOCK(cs_main);
+                auto chainstates{chainman.GetAll()};
+                if (std::any_of(chainstates.begin(), chainstates.end(),
+                                [&chainparams](CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return cs->NeedsRedownload(chainparams); })) {
```
There's no harm in doing this, but I don't think making the pointer const (`CChainstate* const`) is an improvement. In c++, when you pass a pointer, the called function makes a copy of that pointer, so it being const or not communicates nothing to the caller.

Making the pointer point to const data (`const CChainstate*`) on the other hand _is_ useful. It means that the function can't mutate the thing that `cs` is pointing to (and can't be called with a pointer to non-const).

glozow replied:
Ah true 👍 the latter `const` is what I was going for :D

dhruv replied:
Done.

glozow commented on test/functional/p2p_segwit.py:
```diff
@@ -1956,22 +1956,35 @@ def test_non_standard_witness(self):
-            assert_equal(self.nodes[0].getblock(block_hash), self.nodes[2].getblock(block_hash))
-            height -= 1
+        # The upgraded node syncs headers and performs redownload
+        for n in range(2):
+            assert_equal(self.nodes[n].getblockcount(), self.nodes[2].getblockcount())
```
is the assert_equal for `getblockcount()` necessary?

dhruv replied:
You're right. `sync_blocks()` checks for the best block hash as mentioned, so the `getblockcount()` check is really just checking for very unlikely hash collisions. Updated.

dhruv commented:
Addressed comments from @glozow and simplified the functional test a bit - previous code was checking segwit activation on all nodes, but we only need the check on the upgrading node. Ready for further review.

jamesob commented on test/functional/p2p_segwit.py:
```diff
@@ -1956,22 +1956,33 @@ def test_non_standard_witness(self):
-            block_hash = self.nodes[2].getblockhash(height)
-            assert_equal(block_hash, self.nodes[0].getblockhash(height))
-            assert_equal(self.nodes[0].getblock(block_hash), self.nodes[2].getblock(block_hash))
-            height -= 1
```
So based on the removal of this test code (and the `SEGWIT_HEIGHT - 1` assertion above), the implication seems to be we expect users who are upgrading from pre-segwit to remove their datadir and sync to tip from scratch? Based on the assertions here, `-reindex` will not bring them to tip, so what's the point of telling the user to reindex at all? Or is this behavior specific to the functional tests, and in practice `-reindex` *would* bring the upgraded node to tip? (I'll need to review the reindex code; I can't remember what to expect there.)

If we expect users to remove their blocksdir before restarting with `-reindex`, you might consider mentioning that in the error message you've included in `init.cpp`.

jnewbery replied:
Reindexing will result in the node discarding the blocks without witness after segwit activation and redownloading them from peers and validating. See https://github.com/bitcoin/bitcoin/pull/21009#discussion_r576134367. The reason the node doesn't sync to tip here is because in our functional tests, the nodes don't make automatic connections to each other. As soon as the nodes are connected, the upgraded node will resync from its peers.

dhruv replied:
Ah, I should have updated the PR description. Updated to now say:

> This PR introduces NeedsRedownload() which merely checks for insufficiently validated segwit blocks and requests that the user restarts the node with `-reindex`. Reindexing the block files upon restart will make the node rebuild chain state and block index from the `blk*.dat` files on disk. The node won't be able to index the blocks with `BLOCK_OPT_WITNESS`, so they will me missing from the chain and be re-downloaded, with witness data.

Sorry for the confusion. 

As for the `-reindex` code, the call path is: `src/init.cpp::Threadimport()` -> `validation.cpp:CChainState::LoadExternalBlockFile()` -> `validation.cpp:CChainState::AcceptBlock()` -> `validation.cpp:ContextualCheckBlock()`. The last function tries to [validate for witness commitments](https://github.com/bitcoin/bitcoin/blob/4b5659c6b115315c9fd2902b4edd4b960a5e066e/src/validation.cpp#L3418) and is [unable to](https://github.com/bitcoin/bitcoin/blob/4b5659c6b115315c9fd2902b4edd4b960a5e066e/src/validation.cpp#L3436) thereby causing the block to not be accepted into the chain upon a restart with `-reindex`

Using `combine_logs.py` for `test/functional/p2p_segwit.py` you can see lines like:

```
 node2 2021-04-22T14:56:56.543661Z [loadblk] [util/system.h:52] [error] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size
 node2 2021-04-22T14:56:56.544812Z [loadblk] [util/system.h:52] [error] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size
```

jamesob replied:
Ah, thanks all. Makes sense.

jamesob reviewed (commented):
Looks good generally and I'm very much in favor of this change.

glozow reviewed (commented):
utACK d831e711cab83c70bf2ded62fe33f484844e73dd

jnewbery commented:
utACK d831e711cab83c70bf2ded62fe33f484844e73dd

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/pull/21009/commits/d831e711cab83c70bf2ded62fe33f484844e73dd

Will be testing locally today.

jamesob commented:
Built and ran tests locally. Grepped around for lingering references and [found a few](https://github.com/jamesob/bitcoin/blob/d831e711cab83c70bf2ded62fe33f484844e73dd/src/chain.h#L185) in documentation that can be cleaned up after this PR. Snapshot activation docs should also be changed in a follow-up PR but the existing logic there is still necessary: we still need to "fake" `BLOCK_OPT_WITNESS` in the way that we are now to avoid tripping the new `NeedsRedownload()` check.

Thanks for this change @dhruv. Nice work!

laanwj commented:
Cursory code review ACK d831e711cab83c70bf2ded62fe33f484844e73dd. Agree with the direction of the change, thanks for simplifying the logic here.

dhruv commented:
> Grepped around for lingering references and [found a few](https://github.com/jamesob/bitcoin/blob/d831e711cab83c70bf2ded62fe33f484844e73dd/src/chain.h#L185) in documentation that can be cleaned up after this PR

@jamesob I've attempted to update references in #21816
